### PR TITLE
bruno: package from source

### PR DESCRIPTION
**Pull request by lucasew**
## Description of changes
Package bruno from source as now they ship package-lock.json together with the repo.

To fix the package-lock.json issue, that upstream doesn't seem to care, I also integrated the tool mentioned in #261137. Using the patch approach requires manual intervention at each bump and generating the package-lock.json is tricky because the tool that fetches src would actually fetch the patched package-lock.json to fix. It's much simpler to integrate the fix into a derivation and also easier bumps using r-ryantm.

I checked if npm-lockfile-fix creates a reproducible result by running the cold build of `bruno.src` twice and it gave the same hash.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 261139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 261139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 261139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 261139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>bruno</li>
  </ul>
</details>

**Comment by lucasew** [comment]
Upstream seems to not care about the package-lock.json thing.

So I guess we have to regenerate a patch at each bump.

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 261139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>bruno</li>
  </ul>
</details>

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 261139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>bruno</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Is this PR ready to review? Because in the description it says it is waiting for some issue to be closed first.

**Comment by lucasew** [comment]
> Is this PR ready to review? Because in the description it says it is waiting for some issue to be closed first.

I used a tool mentioned in the issue to generate the patch of package-lock.json.

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 261139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>bruno</li>
    <li>npm-lockfile-fix</li>
    <li>npm-lockfile-fix.dist</li>
  </ul>
</details>

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 261139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>bruno</li>
    <li>npm-lockfile-fix</li>
    <li>npm-lockfile-fix.dist</li>
  </ul>
</details>

File: pkgs/by-name/br/bruno/package.nix [hunk 1]
```diff
@@ -1,61 +1,127 @@
...
+, pixman
+, cairo
+, pango
+, npm-lockfile-fix
 }:
 
-stdenv.mkDerivation rec {
+let
+  electron = electron-bin;
+in
```
Comment (SuperSandro2000) [inline pkgs/by-name/br/bruno/package.nix:20]: Whats the reason behind this?

File: pkgs/by-name/br/bruno/package.nix [hunk 2]
```diff
@@ -1,61 +1,127 @@
...
-    hash = "sha256-ptrayWDnRXGUC/mgSnQ/8sIEdey+6uoa3LGBGPQYuY8=";
+  src = fetchFromGitHub {
+    owner = "usebruno";
+    repo = "bruno";
+    rev = "v${version}";
+    hash = "sha256-GgXnsPEUurPHrijf966x5ldp+1lDrgS1iBinU+EkdYU=b";
+
+    postFetch = ''
+      ${lib.getExe npm-lockfile-fix} $out/package-lock.json
+    '';
```
Comment (SuperSandro2000) [inline pkgs/by-name/br/bruno/package.nix:34]: I think we should rather do this in postPatch to make the download easier to reproduce.

Also that this does network requests is not that great. We should probably just apply a patch to get the hashes.

File: pkgs/by-name/br/bruno/package.nix [hunk 3]
```diff
@@ -1,61 +1,127 @@
...
-  postFixup = ''
-    substituteInPlace "$out/share/applications/bruno.desktop" \
-      --replace "/opt/Bruno/bruno" "$out/bin/bruno"
+    mkdir -p $out/opt/bruno $out/bin
+
+    pushd packages/bruno-electron/dist/linux-unpacked
+
+      cp -r locales resources{,.pak} $out/opt/bruno
+
+    popd
```
Comment (SuperSandro2000) [inline pkgs/by-name/br/bruno/package.nix:101]:
```suggestion
    cp -r packages/bruno-electron/dist/linux-unpacked/{locales,resources{,.pak}} $out/opt/bruno
```
Does this work? I am not sure

File: pkgs/by-name/br/bruno/package.nix (hunk 1 above)
Comment (lucasew) [reply pkgs/by-name/br/bruno/package.nix:20]: Only convenience. I was testing with different electron versions at first and settled for the default one.

File: pkgs/by-name/br/bruno/package.nix (hunk 2 above)
Comment (lucasew) [reply pkgs/by-name/br/bruno/package.nix:34]: Build complains about inconsistencies in package-lock.json.

That way the bump script already fixes the lock and updates the hash. As the lock fix is reproducible it's fine.

fetchFromGitHub doesn't have a patchPhase tho

File: pkgs/by-name/br/bruno/package.nix (hunk 3 above)
Comment (lucasew) [reply pkgs/by-name/br/bruno/package.nix:101]: I need to test this tomorrow

Comment (lucasew) [reply pkgs/by-name/br/bruno/package.nix]: It's working. Thank you!

**Comment by yu-re-ka** [comment]
electron-bin is not "the default". That would be just  `electron`.

31 Dec 2023 04:54:27 Lucas Eduardo ***@***.***>:

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 261139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>bruno</li>
    <li>npm-lockfile-fix</li>
    <li>npm-lockfile-fix.dist</li>
  </ul>
</details>

File: pkgs/by-name/br/bruno/package.nix [hunk 4]
```diff
@@ -1,62 +1,123 @@
...
+      comment = "Opensource API Client for Exploring and Testing APIs";
+      categories = [ "Development" ];
+      startupWMClass = "Bruno";
+    })
   ];
 
-  runtimeDependencies = [ (lib.getLib systemd) ];
+  postPatch = ''
+    substituteInPlace scripts/build-electron.sh \
+      --replace 'if [ "$1" == "snap" ]; then' 'exit 0; echo foi; if [ "$1" == "snap" ]; then'
```
Comment (thiagokokada) [inline pkgs/by-name/br/bruno/package.nix:66]: This `echo foi` should probably be removed.

Comment (lucasew) [reply pkgs/by-name/br/bruno/package.nix:66]: I love code review lol

Comment (lucasew) [reply pkgs/by-name/br/bruno/package.nix]: Solved

**Review by thiagokokada — APPROVED** [review]

**Review by thiagokokada — APPROVED** [review]

File: pkgs/by-name/br/bruno/package.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/by-name/br/bruno/package.nix:30]: > As the lock fix is reproducible it's fine.

but we now depend on npmjs.com working to be able to fetch the src, right? If someone has a copy of the archive but the service is no longer available or some download is yanked, then the FOD cannot be produced anymore. A patch file shouldn't have that problem.

Comment (lucasew) [reply pkgs/by-name/br/bruno/package.nix:30]: We will still depend on npm to generate the node_modules to build the package. BTW for end users this bundle will be cached.

**Pull request merged**